File: AFSwinx/src/main/java/com/tomscz/afswinx/rest/connection/ConnectionParser.java
package com.tomscz.afswinx.rest.connection;

import java.util.HashMap;

import com.tomscz.afswinx.common.ParameterMissingException;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import com.tomscz.afswinx.common.Utils;
import com.tomscz.afrest.commons.AFRestUtils;

/**
 * This class parse connection from input file (xml) and build connection from them.
 * 
 * @author Martin Tomasek (martin@toms-cz.com)
 * 
 * @since 1.0.0.
 */
public class ConnectionParser implements XMLParser {

    // Tags which separate connections
    private static final String CONNECTION_TAG = "connection";
    private static final String CONNECTION_ID_ATTRIBUTE = "id";

    // Tags which separate connection type
    private static final String METAMODEL_CONNECTION = "metaModel";
    private static final String DATA_CONNECTION = "data";
    private static final String SEND_CONNECTION = "send";
    
    // Tags which will specify security
    private static final String SECURITY_METHOD = "security-method";
    private static final String USER_NAME = "userName";
    private static final String PASSOWORD = "password";
    

    // Concrete value of connection
    private static final String END_POINT = "endPoint";
    private static final String END_POINT_PARAMETERS = "endPointParameters";
    private static final String PROTOCOL = "protocol";
    private static final String PORT = "port";
    private static final String HEADER_PARAM = "header-param";
    private static final String SECURITY_PARAMS = "security-params";
    private static final String PARAM = "param";
    private static final String VALUE = "value";
    private static final String CONTENT_TYPE = "content-type";
    private static final String ACCEPT_TYPE = "accept-type";
    private static final String HTTP_METHOD = "method";

    private HashMap<String, String> elConnectionData;

    // Flag if EL evaluation will be done
    private boolean doElEvaluation = false;
    // Id of connection to parse
    private String connectionId;

    /**
     * Constructor.
     * 
     * @param id of connection which will be parsed
     * @param elConnectionData data which will be used to replace EL variables.
     */
    public ConnectionParser(String id, HashMap<String, String> elConnectionData) {
        this.connectionId = id;
        if (elConnectionData != null && !elConnectionData.isEmpty()) {
            this.elConnectionData = elConnectionData;
            doElEvaluation = true;
        }
    }

    /**
     * Constructor
     * 
     * @param id of connection will will be parsed
     */
    public ConnectionParser(String id) {
        this.connectionId = id;
    }

    /**
     * This method parse document in which are defined connection.
     * 
     * @return {@link AFSwinxConnectionPack} which holds all connection which are in documents and
     *         which belongs to concrete connection key.
     */
    @SuppressWarnings("unchecked")
    @Override
    public AFSwinxConnectionPack parseDocument(Document documentToParse) throws ParameterMissingException {
        // Prepare connection pack
        AFSwinxConnectionPack connectionPack = new AFSwinxConnectionPack();
        // Find root of document
        NodeList childs = documentToParse.getElementsByTagName(CONNECTION_TAG);
        for (int i = 0; i < childs.getLength(); i++) {
            Node connectionRoot = childs.item(i);
            if (connectionRoot.getNodeType() != Node.ELEMENT_NODE) {
                continue;
            }
            // For each connection check if it's id is id which must be parsed
            Element nodeElement = (Element) connectionRoot;
            if (nodeElement.getAttribute(CONNECTION_ID_ATTRIBUTE).equals(connectionId)) {
                NodeList connectionsTypes = connectionRoot.getChildNodes();
                for (int j = 0; j < connectionsTypes.getLength(); j++) {
                    Node concreteConnection = connectionsTypes.item(j);
                    if (concreteConnection.getNodeType() != Node.ELEMENT_NODE) {
                        continue;
                    }
                    String connectionName = concreteConnection.getNodeName();
                    NodeList nodeProperties = concreteConnection.getChildNodes();
                    // Create new connection
                    AFSwinxConnection connection = new AFSwinxConnection();
                    for (int k = 0; k < nodeProperties.getLength(); k++) {
                        // Set connection properties
                        Node property = nodeProperties.item(k);
                        String nodeName = property.getNodeName();
                        String nodeValue = property.getTextContent();
                        if (nodeName.equals(END_POINT)) {
                            connection.setAddress(evaluateEL(nodeValue));
                        } else if (nodeName.equals(END_POINT_PARAMETERS)) {
                            connection.setParameters(evaluateEL(nodeValue));
                        } else if (nodeName.equals(PROTOCOL)) {
                            connection.setProtocol(evaluateEL(nodeValue));
                        } else if (nodeName.equals(PORT)) {
                        	String port = evaluateEL(nodeValue);
                            if(!port.isEmpty()) {
                                connection.setPort(Utils.convertStringToInteger(evaluateEL(nodeValue)));
                            }
                        } else if (nodeName.equals(HEADER_PARAM)) {
                            parseHeaderParam(connection, property.getChildNodes());
                        } else if (nodeName.equals(HTTP_METHOD)) {
                            String method = evaluateEL(nodeValue);
                            HttpMethod httpMethod =
                                    (HttpMethod) AFRestUtils.getEnumFromString(HttpMethod.class,
                                            method.toLowerCase(), true);
                            connection.setHttpMethod(httpMethod);
                        }
                        else if(nodeName.equals(SECURITY_PARAMS)){
                            parseSecurityParams(connection, property.getChildNodes());
                        }
                    }
                    // Set created connection to connection holder based on connection type
                    if (connectionName.equals(METAMODEL_CONNECTION)) {
                        connection.setHttpMethod(HttpMethod.GET);
                        connectionPack.setMetamodelConnection(connection);
                    } else if (connectionName.equals(DATA_CONNECTION)) {
                        connection.setHttpMethod(HttpMethod.GET);
                        connectionPack.setDataConnection(connection);
                    } else if (connectionName.equals(SEND_CONNECTION)) {
                        if (connection.getHttpMethod() == null) {
                            connection.setHttpMethod(HttpMethod.POST);
                        }
                        connectionPack.setSendConnection(connection);
                    }
                }
            }
        }
        return connectionPack;
    }
    
    /**
     * This method parse security and set it to connection.
     * @param connection current connection
     * @param securityParams node with security options
     */
    private void parseSecurityParams(AFSwinxConnection connection, NodeList securityParams) throws ParameterMissingException {
        ConnectionSecurity security = new ConnectionSecurity();
        for(int i=0;i<securityParams.getLength();i++){
            Node node = securityParams.item(i);
            String nodeName = node.getNodeName();
            String nodeValue = evaluateEL(node.getTextContent()).toLowerCase();
            if(nodeName.equals(SECURITY_METHOD)){
                if(nodeValue.equals(SecurityMethod.BASIC.toString())){
                    security.setMethod(SecurityMethod.BASIC);
                }
            }
            else if(nodeName.equals(USER_NAME)){
                security.setUserName(nodeValue);
            }
            else if(nodeName.equals(PASSOWORD)){
                security.setPassword(nodeValue);
            }
        }
        connection.setSecurity(security);
    }

    private void parseHeaderParam(AFSwinxConnection connection, NodeList headerParam) throws ParameterMissingException {
        String key = "";
        String value = "";
        for (int i = 0; i < headerParam.getLength(); i++) {
            Node concreteParam = headerParam.item(i);
            if (concreteParam.getNodeType() != Node.ELEMENT_NODE) {
                continue;
            }
            String nodeName = concreteParam.getNodeName();
            String nodeValue = concreteParam.getTextContent();
            if(nodeName.equals(PARAM)){
                key = evaluateEL(nodeValue);
            }
            else if(nodeName.equals(VALUE)){
                value = evaluateEL(nodeValue);
            }
        }
        // Parse content type and accept type separately
        if (key.equals(CONTENT_TYPE)) {
            connection.setContentType((HeaderType) AFRestUtils.getEnumFromString(
                    HeaderType.class, value, true));
        } else if (key.equals(ACCEPT_TYPE)) {
            connection.setAcceptedType((HeaderType) AFRestUtils.getEnumFromString(
                    HeaderType.class, value, true));
        }
        else{
            connection.addHeaderParam(key, value);
        }
    }

    /**
     * This method determine if EL parse will be done or not.
     * 
     * @param nodeValue value of node in which will be used EL parser
     * @return value of node after execute EL parser on it. If parser should not be execute then
     *         return received value without modification
     */
    private String evaluateEL(String nodeValue) throws ParameterMissingException {
        if (doElEvaluation && nodeValue != null && !nodeValue.isEmpty()) {
            return Utils.evaluateElExpression(nodeValue, elConnectionData);
        }
        return nodeValue;
    }

}

File: AFMobile/AFWindowsPhone/AFWindowsPhone/parsers/ConnectionParser.cs
using AFWindowsPhone.rest.connection;
using AFWindowsPhone.utils;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Linq;
using Windows.Data.Xml.Dom;
using Windows.Web.Http;

namespace AFWindowsPhone.parsers
{
    class ConnectionParser : XMLParser
    {

        // Tags which separate connections
        private const String CONNECTION_TAG = "connection";
        private const String CONNECTION_ID_ATTRIBUTE = "id";

        // Tags which separate connection type
        private const String METAMODEL_CONNECTION = "metaModel";
        private const String DATA_CONNECTION = "data";
        private const String SEND_CONNECTION = "send";
        private const String REMOVE_CONNECTION = "remove";

        // Tags which will specify security
        private const String SECURITY_METHOD = "security-method";
        private const String USER_NAME = "userName";
        private const String PASSOWORD = "password";


        // Concrete value of connection
        private const String END_POINT = "endPoint";
        private const String END_POINT_PARAMETERS = "endPointParameters";
        private const String PROTOCOL = "protocol";
        private const String PORT = "port";
        private const String HEADER_PARAM = "header-param";
        private const String SECURITY_PARAMS = "security-params";
        private const String PARAM = "param";
        private const String VALUE = "value";
        private const String CONTENT_TYPE = "content-type";
        private const String ACCEPT_TYPE = "accept-type";
        private const String HTTP_METHOD = "method";

        private Dictionary<String, String> elConnectionData;

        // Flag if EL evaluation will be done
        private bool doElEvaluation = false;
        // Id of connection to parse
        private String connectionId;

        public ConnectionParser(String id, Dictionary<String, String> elConnectionData)
        {
            this.connectionId = id;
            if (elConnectionData != null && elConnectionData.Count != 0)
            {
                this.elConnectionData = elConnectionData;
                doElEvaluation = true;
            }
        }


        public ConnectionParser(String id)
        {
            this.connectionId = id;
        }

        public AFSwinxConnectionPack parseDocument(XmlDocument documentToParse)
        {
            // Prepare connection pack
            AFSwinxConnectionPack connectionPack = new AFSwinxConnectionPack();
            // Find root of document
            XmlNodeList childs = documentToParse.GetElementsByTagName(CONNECTION_TAG);
            foreach (IXmlNode connectionRoot in childs)
            {
                if (connectionRoot.NodeType != NodeType.ElementNode)
                {
                    continue;
                }
                // For each connection check if it's id is id which must be parsed
                XmlElement nodeElement = (XmlElement)connectionRoot;
                if (nodeElement.GetAttribute(CONNECTION_ID_ATTRIBUTE).Equals(connectionId))
                {
                    XmlNodeList connectionsTypes = connectionRoot.ChildNodes;
                    for (uint j = 0; j < connectionsTypes.Length; j++)
                    {
                        IXmlNode concreteConnection = connectionsTypes.Item(j);
                        if (concreteConnection.NodeType != NodeType.ElementNode)
                        {
                            continue;
                        }
                        String connectionName = concreteConnection.NodeName;
                        XmlNodeList nodeProperties = concreteConnection.ChildNodes;
                        // Create new connection
                        AFSwinxConnection connection = new AFSwinxConnection();
                        for (uint k = 0; k < nodeProperties.Length; k++)
                        {
                            // Set connection properties
                            IXmlNode property = nodeProperties.Item(k);
                            String nodeName = property.NodeName;
                            String nodeValue = property.InnerText;
                            if (nodeName.Equals(END_POINT))
                            {
                                connection.setAddress(evaluateEL(nodeValue));
                            }
                            else if (nodeName.Equals(END_POINT_PARAMETERS))
                            {
                                connection.setParameters(evaluateEL(nodeValue));
                            }
                            else if (nodeName.Equals(PROTOCOL))
                            {
                                connection.setProtocol(evaluateEL(nodeValue));
                            }
                            else if (nodeName.Equals(PORT))
                            {
                                String port = evaluateEL(nodeValue);
                                if (!String.IsNullOrEmpty(port))
                                {
                                    connection.setPort(Convert.ToInt32(evaluateEL(nodeValue)));
                                }
                            }
                            else if (nodeName.Equals(HEADER_PARAM))
                            {
                                parseHeaderParam(connection, property.ChildNodes);
                            }
                            else if (nodeName.Equals(HTTP_METHOD))
                            {
                                String method = evaluateEL(nodeValue);
                                if (method.ToLower().Equals("get"))
                                {
                                    connection.setHttpMethod(HttpMethod.Get);
                                }
                                else if (method.ToLower().Equals("post"))
                                {
                                    connection.setHttpMethod(HttpMethod.Post);
                                }
                                else if (method.ToLower().Equals("put"))
                                {
                                    connection.setHttpMethod(HttpMethod.Put);
                                }
                                else if (method.ToLower().Equals("delete"))
                                {
                                    connection.setHttpMethod(HttpMethod.Delete);
                                }
                                else
                                {
                                    //method not supported
                                }
                                //get method
                            }
                            else if (nodeName.Equals(SECURITY_PARAMS))
                            {
                                parseSecurityParams(connection, property.ChildNodes);
                            }
                        }
                        // Set created connection to connection holder based on connection type
                        if (connectionName.Equals(METAMODEL_CONNECTION))
                        {
                            connection.setHttpMethod(HttpMethod.Get);
                            connectionPack.setMetamodelConnection(connection);
                        }
                        else if (connectionName.Equals(DATA_CONNECTION))
                        {
                            connection.setHttpMethod(HttpMethod.Get);
                            connectionPack.setDataConnection(connection);
                        }
                        else if (connectionName.Equals(SEND_CONNECTION))
                        {
                            if (connection.getHttpMethod() == null)
                            {
                                connection.setHttpMethod(HttpMethod.Post);
                            }
                            connectionPack.setSendConnection(connection);
                        }
                        else if (connectionName.Equals(REMOVE_CONNECTION))
                        {
                            if (connection.getHttpMethod() == null)
                            {
                                connection.setHttpMethod(HttpMethod.Delete);
                            }
                        }
                    }
                }
            }
            return connectionPack;
        }

        private void parseSecurityParams(AFSwinxConnection connection, XmlNodeList securityParams)
        {
            ConnectionSecurity security = new ConnectionSecurity();
            for (uint i = 0; i < securityParams.Length; i++)
            {
                IXmlNode node = securityParams.Item(i);
                String nodeName = node.NodeName;
                String nodeValue = evaluateEL(node.InnerText).ToLower();
                if (nodeName.Equals(SECURITY_METHOD))
                {
                    if (nodeValue.Equals(SecurityMethod.BASIC.toString()))
                    {
                        security.setMethod(SecurityMethod.BASIC);
                    }
                }
                else if (nodeName.Equals(USER_NAME))
                {
                    security.setUserName(nodeValue);
                }
                else if (nodeName.Equals(PASSOWORD))
                {
                    security.setPassword(nodeValue);
                }
            }
            connection.setSecurity(security);
        }

        private void parseHeaderParam(AFSwinxConnection connection, XmlNodeList headerParam)
        {
            String key = "";
            String value = "";
            for (uint i = 0; i < headerParam.Length; i++)
            {
                IXmlNode concreteParam = headerParam.Item(i);
                if (concreteParam.NodeType != NodeType.ElementNode)
                {
                    continue;
                }
                String nodeName = concreteParam.NodeName;
                String nodeValue = concreteParam.InnerText;
                if (nodeName.Equals(PARAM))
                {
                    key = evaluateEL(nodeValue);
                }
                else if (nodeName.Equals(VALUE))
                {
                    value = evaluateEL(nodeValue);
                }
            }
            // Parse content type and accept type separately
            if (key.Equals(CONTENT_TYPE))
            {
                /* connection.setContentType((HeaderType)Utils.getEnumFromString(
                         HeaderType.class, value, true));*/
            }
            else if (key.Equals(ACCEPT_TYPE))
            {
                /*connection.setAcceptedType((HeaderType) Utils.getEnumFromString(
                        HeaderType.class, value, true));*/
            }
            else {
                connection.addHeaderParam(key, value);
            }
        }

        private String evaluateEL(String nodeValue)
        {
            if (doElEvaluation && !String.IsNullOrEmpty(nodeValue))
            {
                return evaluateElExpression(nodeValue, elConnectionData);
            }
            return nodeValue;
        }

        private String evaluateElExpression(String expressionToEvaluate,
                                              Dictionary<String, String> parameters)
        {
            // To chaining string use string builder
            StringBuilder replacedValue = new StringBuilder();
            // Split expression by #{ it gives you strings between and after value which should be
            // replaced
            String[] values = expressionToEvaluate.Split(new []{ "#{" }, StringSplitOptions.None);
            bool firstCycle = true;
            foreach (String value in values)
            {
                // in first cycle add everything because split is done by #{ which means that before
                // first char sequence #{ is plain text
                if (firstCycle)
                {
                    replacedValue.Append(value.Substring(0, value.Length));
                    firstCycle = false;
                    continue;
                }
                // This are values behind string which will be replaced. in first position is string to
                // replaced
                String[] valuesBehind = value.Split(new []{"}"}, StringSplitOptions.None);
                // Find replaced value and append it
                String elValue = parameters[valuesBehind[0].Substring(0, valuesBehind[0].Length)];
                replacedValue.Append(elValue);
                // If some values left - this means that there is more } brackets ex: #{value}/a}/a}
                // then append them too
                for (int i = 1; i < valuesBehind.Length; i++)
                {
                    if (String.IsNullOrEmpty(valuesBehind[i]))
                    {
                        continue;
                    }
                    replacedValue.Append(valuesBehind[i]);
                    char firstChar = valuesBehind[i][0];
                    //Because split was done by } then it should not be there, then add it if left brackets
                    if (firstChar == '{')
                    {
                        replacedValue.Append("}");
                    }

                }
            }
            return replacedValue.ToString();
        }

    }
}
